### ISSUE TITLE
Token 저장 위치에 따른 보안 issue 

### ISSUE THREAD

**Issue by SoYoung210** [Yapp-Web5/Druwa-WEB#26]
안녕하세요, yapp 4팀에서 FE 개발을 담당하고 있는 이소영 이라고 합니다. 
최근 OAuth2.0 을 활용한 로그인을 구현하며 토큰에 대해 알게된 점을 공유하고자 합니다. 

https://github.com/Yapp-Web5/Druwa-WEB/blob/5a0ca57f13ab2e2d95f4f96716f1adf255858d24/src/sagas/userSaga.js#L6

[token 저장 위치관련 article](http://lazyhoneyant.tistory.com/7)

해당 내용을 보면, `localStorage` 를 활용한 토큰 탈취는 매우 쉽습니다.
[XSS(cross-site scripting)](http://4rgos.tistory.com/1) 공격에 취약합니다.

구체적으로 어떠한 방식으로 인증을 진행하시는지는 알 수 없어, 코멘트를 다는 것이 조금 조심스럽지만 일단 저희가 생각하는 방식은 아래와 같습니다. 
> token-type : Bearer, 백단에서는 jsonwebtoken 라이브러리를 사용하며 OAuth2.0 기반 입니다.  

1. 최초 로그인 시 client에서 access_token 획득 후  BE에서 `jwt.sign` 으로 정보 셋팅. (벨로퍼트님 링크와 유사한 과정)
2. 이후 접근 시 미들웨어에서 검증

```javascript
//토큰 찾기
const token = req.cookies.token || req.headers['x-access-token'] 
```
2-1. `jwt.verify` 사용

생각하고 계신 방식이 구글로그인을 적용하고 있는 저희와는 많이 다를 수 있지만 배포 후 업데이트 작업 하실 때 고려하시면 좋을 것 같아 이슈로 남깁니다. 

완벽히 코드를 보지 못해 섣불리 보안에 이슈가 있다, 라고 말씀드리는 것이 조심스럽습니다.
만약, 제가 여러가지를 간과하고 본 이슈를 남긴것이라면 짤막한 코멘트와 함께  바로 `Close` 처리 해주시면 감사하겠습니다. 
과정도 짤막하게 달았지만 혹시 더 궁금하시다면 댓글로 남겨주세요. 
> 아직 Repo에 로그인을 구현해두지 않아서,,,,, 레포링크는 남기지 않았습니다. 

-----------------
additional link
[Node.js / JWT Express.js 서버에서 JWT 기반 회원인증 ](https://velopert.com/2448)
[JWT Claim Set](https://blog.outsider.ne.kr/1160)
[implement-jwt](https://swalloow.github.io/implement-jwt)

**Comment by nayunhwan** [comment]
네, 이 문제에 대해서는 저번에 소영님과 얘기를 나누다가, localStorage에 저장하는 것이 안전하지 않다는 것을 확인 후 고쳐야하는 부분 중 하나로 생각하고 있습니다.

현재 저희는 token을 개인화 된 사용자 인증보다는, 단순히 동일 사람 판별 정도로만 사용하고 있어, 아직까지는 토큰 탈취로 인한 크리티컬한 side effect는 없을 것이라 판단 됩니다.

물론, side effect가 없다고 해서, 이 문제점을 개선할 여지가 없다는 것은 아니며, 동일 유저 판단 이외에 다른 `permission`에 따른 다른 action이 필요로 하게 된다면, 점차적으로 JWT와 같은 좀 더 안전한 토큰 체계를 사용할 계획입니다.

좋은 Feedback 감사합니다.

**Comment by SoYoung210** [comment]
답변 감사드립니다.
버그 없는 좋은 하루 보내세요👻

**Comment by SoYoung210** [comment]
답변 감사드립니다.
버그 없는 좋은 하루 보내세요👻

**Comment by nayunhwan** [comment]
@SoYoung210 
앞으로 해결해야할 문제이므로, Re-open 하겠습니다.
